File: formula-cs/Formula/ResolvedRollValue.cs
namespace Formula;

public class ResolvedRollValue : ResolvedValue
{
    private readonly int _count;
    private readonly int _sides;

    public ResolvedRollValue(int count, int sides)
    {
        _count = count;
        _sides = sides;
    }

    public override string AsText()
    {
        return _count + "d" + _sides;
    }

    public override int AsNumber()
    {
        return (int)AsDecimal();
    }

    public override double AsDecimal()
    {
        return (_count * (_sides + 1)) / 2.0;
    }

    public override bool AsBoolean()
    {
        return _count > 0 && _sides > 0;
    }

    public override bool Equals(object? obj)
    {
        return obj is ResolvedValue { HasValue: true } other && Math.Abs(AsDecimal() - other.AsDecimal()) < 0.01;
    }

    public override int GetHashCode()
    {
        return _count.GetHashCode() + _sides.GetHashCode();
    }

    public override string ToString()
    {
        return AsText();
    }
}

File: formula-js/src/ResolvedRollValue.ts
import {ResolvedValue} from "./ResolvedValue";

export class ResolvedRollValue implements ResolvedValue {
  constructor(private readonly count: number,
              private readonly sides: number) {
  }

  asText(): string {
    return this.count + 'd' + this.sides;
  }

  asNumber(): number {
    return (this.count * (this.sides + 1)) / 2.0;
  }

  asBoolean(): boolean {
    return this.count > 0 && this.sides > 0;
  }

  asList(): ResolvedValue[] {
    return [this];
  }

  equals(other: ResolvedValue): boolean {
    return other.hasValue() && this.asText() === other.asText();
  }

  toString(): string {
    return this.asText();
  }

  map(fn: (value: ResolvedValue) => ResolvedValue): ResolvedValue {
    return fn(this);
  }

  hasValue(): boolean {
    return true;
  }
}